File: mobile-engage/src/main/java/com/emarsys/mobileengage/client/DefaultClientServiceInternal.java
package com.emarsys.mobileengage.client;

import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.request.RequestManager;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.mobileengage.request.MobileEngageRequestModelFactory;

public class DefaultClientServiceInternal implements ClientServiceInternal {
    private final MobileEngageRequestModelFactory requestModelFactory;
    private final RequestManager requestManager;

    public DefaultClientServiceInternal(RequestManager requestManager, MobileEngageRequestModelFactory requestModelFactory) {
        this.requestModelFactory = requestModelFactory;
        this.requestManager = requestManager;
    }

    @Override
    public void trackDeviceInfo(CompletionListener completionListener) {
        RequestModel requestModel = requestModelFactory.createTrackDeviceInfoRequest();

        requestManager.submit(requestModel, completionListener);
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/client/DefaultClientServiceInternal.kt
package com.emarsys.mobileengage.client

import com.emarsys.core.api.result.CompletionListener
import com.emarsys.core.request.RequestManager
import com.emarsys.mobileengage.request.MobileEngageRequestModelFactory

class DefaultClientServiceInternal(
    private val requestManager: RequestManager,
    private val requestModelFactory: MobileEngageRequestModelFactory
) : ClientServiceInternal {
    override fun trackDeviceInfo(completionListener: CompletionListener?) {
        try {
            val requestModel = requestModelFactory.createTrackDeviceInfoRequest()
            requestManager.submit(requestModel, completionListener)
        } catch (e: IllegalArgumentException) {
            completionListener?.onCompleted(e)
        }
    }
}